consensus/safety-rules/src/safety_rules.rs

```diff
@@ -90,6 +97,28 @@ impl<T: Payload> TSafetyRules<T> for SafetyRules<T> {
         ))
     }
 
+    fn initialize(&mut self, proof: &ValidatorChangeProof) -> Result<(), Error> {
+        let waypoint = self.persistent_storage.waypoint()?;
```

Comment:
this looks too much, sth like this should work
```
let verifier = VerifierType::Waypoint(waypoint);
let ledger_info = match proof.verify(&verifier).map_err(|_| Error::InvalidValidatorChangeProof)?;
self.valdiator_verifier = Some(ledger_info.next_validator_set().into());
```
if you make self.verifier: VerifierType it can become
```
let ledger_info = match proof.verify(&self.verifier).map_err(|_| Error::InvalidValidatorChangeProof)?;
self.verifier = VerifierType::TrustedVerifier(EpochInfo {
  epoch: ledger_info.epoch() + 1,
  verifier: Arc::new(ledger_info.next_validator_set().into()),
});
```

Reply:
interesting worth exploring, though this does indeed work as is. the above may be simpler at least for now... I haven't scoped out the next step where we receive an end of epoch li but I suspect a similar upgrade will work.

Reply:
it's probably going to be similar to how consensus does the upgrade which is above